File: src/roar_net_api/algorithms/grasp.py
# SPDX-FileCopyrightText: © 2025 Authors of the roar-net-api-py project <https://github.com/roar-net/roar-net-api-py/blob/main/AUTHORS>
#
# SPDX-License-Identifier: Apache-2.0

import random
from collections.abc import Callable
from logging import getLogger
from operator import itemgetter
from time import perf_counter
from typing import Optional, Protocol, TypeVar, Union, cast

from ..operations import (
    SupportsApplyMove,
    SupportsConstructionNeighbourhood,
    SupportsCopySolution,
    SupportsEmptySolution,
    SupportsLowerBoundIncrement,
    SupportsMoves,
    SupportsObjectiveValue,
)

log = getLogger(__name__)


class _Solution(SupportsObjectiveValue, SupportsCopySolution, Protocol): ...


_TSolution = TypeVar("_TSolution", bound=_Solution)


class _Move(SupportsLowerBoundIncrement[_TSolution], SupportsApplyMove[_TSolution], Protocol): ...


class _Neighbourhood(SupportsMoves[_TSolution, _Move[_TSolution]], Protocol): ...


class _Problem(
    SupportsConstructionNeighbourhood[_Neighbourhood[_TSolution]], SupportsEmptySolution[_TSolution], Protocol
): ...


LocalSearchFunc = Callable[[_Problem[_TSolution], _TSolution], _TSolution]


def grasp(
    problem: _Problem[_TSolution],
    budget: float,
    solution: Optional[_TSolution] = None,
    alpha: float = 0.1,
    local_search: Optional[LocalSearchFunc[_TSolution]] = None,
) -> _TSolution:
    start = perf_counter()

    neigh = problem.construction_neighbourhood()

    if solution is None:
        solution = problem.empty_solution()

    best = solution
    bestobj = solution.objective_value()

    while perf_counter() - start < budget:
        s = solution.copy_solution()
        b = None
        bobj = None

        cl = _valid_moves_and_increments(neigh, s)
        while len(cl) != 0:
            cmin = min(cl, key=itemgetter(1))[1]
            cmax = max(cl, key=itemgetter(1))[1]
            thresh = cmin + alpha * (cmax - cmin)
            rcl = [m for m, decr in cl if decr <= thresh]
            m = random.choice(rcl)
            s = m.apply_move(s)
            obj = s.objective_value()
            if obj is not None and (bobj is None or obj < bobj):
                b = s.copy_solution()
                bobj = b.objective_value()
            cl = _valid_moves_and_increments(neigh, s)
        if b is not None:
            if local_search is not None:
                b = local_search(problem, b)
                # The following assumes that local_search returns a better or equal objective value
                bobj = cast(Union[int, float], b.objective_value())
            bobj = cast(Union[int, float], bobj)
            if bestobj is None or bobj < bestobj:
                log.info(f"Best solution: {bobj}")
                best = b
                bestobj = bobj
    return best


def _valid_moves_and_increments(
    neigh: _Neighbourhood[_TSolution], solution: _TSolution
) -> list[tuple[_Move[_TSolution], Union[int, float]]]:
    res: list[tuple[_Move[_TSolution], Union[int, float]]] = []
    for m in neigh.moves(solution):
        incr = m.lower_bound_increment(solution)
        if incr is not None:
            res.append((m, incr))
    return res

File: julia/src/algorithms/grasp.jl
"""
    grasp(problem::Problem, budget::Real; solution::Union{Nothing, Solution}=nothing,
          alpha::Real=0.1, local_search::Union{Nothing, Function}=nothing) -> solution::Solution

Greedy Randomized Adaptive Search Procedure: build solutions using a
restricted candidate list controlled by `alpha`, then optionally improve
with `local_search`. Runs for `budget` seconds.
"""
function grasp(problem::Problem, budget::Real;
        solution::Union{Nothing, Solution}=nothing,
        alpha::Real=0.1,
        local_search::Union{Nothing, Function}=nothing)::Solution

    start = time()

    neigh = construction_neighbourhood(problem)

    if solution === nothing
        solution = empty_solution(problem)
    end

    best = solution
    best_obj = objective_value(solution)

    while time() - start < budget
        s = copy_solution(solution)
        b::Union{Nothing, Solution} = nothing
        b_obj::Union{Nothing, <:Real} = nothing

        cl = _gr_moves_and_increments(neigh, s)
        while !isempty(cl)
            cmin = minimum(m -> m[2], cl)
            cmax = maximum(m -> m[2], cl)
            thresh = cmin + alpha * (cmax - cmin)
            rcl = [m[1] for m in cl if m[2] <= thresh]
            mv = rand(rcl)
            s = apply_move(mv, s)
            obj = objective_value(s)
            if obj !== nothing && (b_obj === nothing || obj < b_obj)
                b = copy_solution(s)
                b_obj = objective_value(b)
            end
            cl = _gr_moves_and_increments(neigh, s)
        end

        if b !== nothing
            if local_search !== nothing
                b = local_search(problem, b)
                b_obj = objective_value(b)
            end
            if best_obj === nothing || b_obj < best_obj
                @info "Best solution: $b_obj"
                best = b
                best_obj = b_obj
            end
        end
    end

    return best
end

function _gr_moves_and_increments(neigh::Neighbourhood, solution::Solution)::Vector{Tuple{Any, <:Real}}
    result::Vector{Tuple{Any, <:Real}} = Tuple{Any, <:Real}[]
    for mv in moves(neigh, solution)
        incr = lower_bound_increment(mv, solution)
        if incr !== nothing
            push!(result, (mv, incr))
        end
    end
    return result
end
